lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        return createDomain(type, hasNullValue, ranges);
+    }
+
+    private boolean isCorruptedColumnIndex(ColumnIndex columnIndex)
+    {
+        if (columnIndex.getMaxValues() == null || columnIndex.getMinValues() == null ||
+                columnIndex.getNullCounts() == null || columnIndex.getNullPages() == null) {
+            return true;
+        }
+
+        if (columnIndex.getMaxValues().size() != columnIndex.getMinValues().size() ||
+                columnIndex.getMaxValues().size() != columnIndex.getNullPages().size() ||
+                columnIndex.getMaxValues().size() != columnIndex.getNullCounts().size()) {
+            return true;
+        }
+
+        return false;
+    }
+
+    // Caller should verify isCorruptedColumnIndex is false first
+    private boolean isEmptyColumnIndex(ColumnIndex columnIndex)
+    {
+        return columnIndex.getMaxValues().size() == 0;
+    }
+
+    public FilterPredicate convertToParquetUdp()
+    {
+        FilterPredicate filter = null;
+
+        // TODO: It could be a Presto bug that we don't see effectivePredicate.getDomains().get() has more than 1 domain.
+        //  For example, the 'where c1=3 or c1=10002' clause should have two domains but it has none
+        //  Todo: we assume the relation cross domains are 'or'
+        for (RichColumnDescriptor column : columns) {
+            Domain domain = effectivePredicate.getDomains().get().get(column);
+            if (domain == null || domain.isNone()) {
+                continue;
+            }
+
+            if (domain.isAll()) {
+                continue;
+            }
+
+            FilterPredicate columnFilter = FilterApi.userDefined(FilterApi.intColumn(ColumnPath.get(column.getPath()).toDotString()), new DomainUserDefinedPredicate(domain));
+            if (filter == null) {
+                filter = columnFilter;
+            }
+            else {
+                filter = FilterApi.or(filter, columnFilter);
+            }
+        }
+
+        return filter;
+    }
+
+    /**
+     * This class implements methods defined in UserDefinedPredicate based on the page statistic and tuple domain(for a column).
+     */
+    static class DomainUserDefinedPredicate<T extends Comparable<T>>
+            extends UserDefinedPredicate<T>
+            implements Serializable
+    {
+        private Domain columnDomain;
+
+        DomainUserDefinedPredicate(Domain domain)
+        {
+            this.columnDomain = domain;
+        }
+
+        @Override
+        public boolean keep(T value)
+        {
+            if (value == null && !columnDomain.isNullAllowed()) {
+                return false;
+            }
+
+            return true;
+        }
+
+        @Override
+        public boolean canDrop(org.apache.parquet.filter2.predicate.Statistics<T> statistic)
+        {
+            if (statistic == null) {
+                return false;
+            }
+            else {
+                if (statistic.getMin() instanceof Integer) {
```

Comment:
As with `createDomain`, this `if`/`else` chain should be made more-generic. Unlike `createDomain`, we can't add a parameter to this method.

Instead, I might create a separate method like `<T> ParquetRangeStatistics<? super T> getRangeStatistics(T min, T max)`, which can be used both here and in `createDomain`.

Also, this particular `else` and `if` should be on the same line.

Reply:
Fixed the `else`/`if`, but the creation of the ParquetRangeStatistics is different than in the previous case and the processing is slightly different in each case. I don't think there's much that can be done to clean it up more.

Reply:
I think the per-type processing could be split into a special class or method, similar to the use of `ParquetXStatistics` in `canDrop`, but to make it clean, it would probably need to be used more widely in the PR, which might not be worth the effort.